### ISSUE TITLE
Templated infinitely recursive function does not fail to compile

### ISSUE THREAD

**Issue by alexkaratarakis** [llvm/llvm-project#141686]
Min repro:
```C++
template <typename R1>
static R1 self_recursion_template(const R1& input)
{
    return self_recursion_template<R1>(input);
}

static int self_recursion_non_template(const int& input)
{
    return self_recursion_non_template(input);
}

int main()
{
    int input{};

    const int result = self_recursion_template(input);
    // const int result = self_recursion_non_template(input);  // Properly fails to compile
    (void)result;
}

```

Expected results: compilation fails for the templated version of infinite self-recursion like it does for the non-templated version of infinite self-recursion.

Actual results: compiles and then stack overflows when trying to run it.

Additionally, `msvc` rejects the code.

Demo: https://godbolt.org/z/eEq7PT771

**Comment by zwuis** [comment]
I think Clang is correct here. But a warning will be emitted if you add command line option `-Winfinite-recursion`. `-Wall` also enables this warning.

**Comment by AaronBallman** [comment]
> I think Clang is correct here. But a warning will be emitted if you add command line option `-Winfinite-recursion`. `-Wall` also enables this warning.

That warning isn't firing for the templated code though: https://godbolt.org/z/WYoo8Y45b

We cannot fail compilation for this scenario (unless the user passed `-Werror`), but I think we should be able to issue the diagnostic for the templated function after instantiation.

**Comment by AaronBallman** [comment]
> ... where DoStuff<0, 0> does indeed recursively call itself, and is indeed instantiated by DoStuff<0, 1>, but the call to it is never reached.

We're already in an analysis-based warning, so it seems like we should have what we need: we could note that we want to issue the diagnostic there, and once we get to the end of the analysis for the function, check whether that location was ever actually reached before issuing the diagnostic.

I wasn't thinking we'd need to modify `DiagRuntimeBehavior()`, I was thinking we'd be basically re-using that logic (but without a second pass over the CFG). But I suspect it would still be somewhat involved because I think we'd have to refactor the reachability analysis to be more reusable.

**Comment by zygoloid** [comment]
> > ... where DoStuff<0, 0> does indeed recursively call itself, and is indeed instantiated by DoStuff<0, 1>, but the call to it is never reached.
> 
> We're already in an analysis-based warning, so it seems like we should have what we need: we could note that we want to issue the diagnostic there, and once we get to the end of the analysis for the function, check whether that location was ever actually reached before issuing the diagnostic.

I think we may be talking about different things. `DoStuff<0, 0>` does indeed recursively call itself through a reachable call. So if we only check whether the recursive call location is reached from the entry of the function (which we already do -- the diagnostic checks whether *every* path from the entry of the function reaches a recursive call), we would produce a false positive diagnostic. The reason it's a false positive is not something that can be seen by analyzing `DoStuff<0, 0>` itself -- it's that `DoStuff<0, 0>` is only ever called from within unreachable code (in `DoStuff<0, 1>`).

**Comment by Alcaro** [comment]
One can also consider how easy it'd be to to fix the code to not emit warnings. For the given example, that'd be a simple if constexpr.

The current behavior was fine when implemented (the linked commit is 2013, if constexpr is C++17), but I'd say it's time to reconsider.
